tests/config_sources/vault_test.py

```diff
@@ -272,3 +272,81 @@ def test_vault_kv_poll_refetch():
             assert wait_for(p(has_datapoint, backend, dimensions={"env": "prod"}))
 
             assert "secret/metadata/app" in audit_read_paths(get_audit_events())
+
+
+READ_ALL_SECRETS_POLICY = """
+path "secret/*" {
+  policy = "read"
+}
+"""
+
+
+def run_vault_iam_test(iam_config):
+    """
+    Test that Vault will authenticate a user with IAM crednetials and provide a
+    token
+    """
+    with run_vault() as [vault_client, _], run_service("awsapi") as aws_cont:
+        aws_ip = container_ip(aws_cont)
+        assert wait_for(p(tcp_socket_open, aws_ip, 8080), 30)
+        vault_client.sys.enable_auth_method("aws")
+
+        vault_client.write(
+            "auth/aws/config/client",
+            **{
+                "iam_endpoint": f"http://{aws_ip}:8080/iam",
+                "sts_endpoint": f"http://{aws_ip}:8080/sts",
+                "ec2_endpoint": f"http://{aws_ip}:8080/ec2",
+                "secret_key": "MY_SECRET_KEY",
+                "access_key": "MY_ACCESS_KEY",
+            },
+        )
+
+        vault_client.sys.create_or_update_policy(name="dev", policy=READ_ALL_SECRETS_POLICY)
+
+        vault_client.write(
+            "auth/aws/role/dev-role-iam",
+            **{"auth_type": "iam", "policies": "dev", "bound_iam_principal_arn": "arn:aws:iam::0123456789:*"},
+        )
+
+        vault_client.write("secret/data/app", data={"env": "dev"})
+
+        with run_agent(
+            dedent(
+                f"""
+            intervalSeconds: 1
+            globalDimensions:
+               env: {{"#from": "vault:secret/data/app[data.env]"}}
+            configSources:
+              vault:
+                vaultAddr: {vault_client.url}
+                authMethod: iam
+                iam: {json.dumps(iam_config)}
+            monitors:
+             - type: collectd/uptime
+        """
+            )
+        ) as [backend, _, _]:
+            assert wait_for(p(has_datapoint, backend, dimensions={"env": "dev"}), timeout_seconds=10)
+
+
+def test_vault_iam_token_with_explicit_credentials():
+    run_vault_iam_test(
+        {"awsAccessKeyId": "MY_ACCESS_KEY", "awsSecretAccessKey": "MY_SECRET_KEY", "role": "dev-role-iam"}
+    )
+
+
+def test_vault_iam_token_with_env_credentials():
+    old_access_key = os.environ.get("AWS_ACCESS_KEY_ID")
+    os.environ["AWS_ACCESS_KEY_ID"] = "MY_ACCESS_KEY"
```

Comment:
I haven't tested it but I think this is a little cleaner:

```python
with monkeypatch.context() as m:
   m.setenv("AWS_ACCESS_KEY_ID", "MY_ACCESS_KEY")
   m.setenv("AWS_SECRET_ACCESS_KEY", "MY_SECRET_KEY")

   ...
```

Reply:
Nice, I was looking for something like this and couldn't find it.  Will update.